README.txt
LuaUnit v3.0 by KwirkyJ

OLD README:

		luaunit.lua  by Philippe Fremy

Luaunit is a testing framework for Lua, in the spirit of Extrem Programming.

Luaunit is derived from the initial work of Ryu Gwang but has evolved a lot
from the original code as my understanding of lua progressed.

Luaunit should work on all platforms supported by lua. It was tested on
Windows XP and Gentoo Linux.

Luaunit is used extensively in yzis (www.yzis.org), a vi clone, in order to
test the lua binding of the editor.

You can download luaunit from:
http://luaforge.net/projects/luaunit/


History:
========

v1.2: first public release

v1.3: ported to lua 5.1

v2.0: many usability and helper functions added. See the top of the source for all the details.



CHANGELOG.txt

[3.0]
- Changes made by KwirkyJ
- Updated compatability for Lua 5.2+ (notably unpack and loadstring).
- Major overhaul of inner workings; 
      most previous code will work without change but some involved scripts 
      (playing with LuaUnit inner variables,e.g.) may fail.
- Added assertNot and assert_not
- Added assertAlmostEquals and assert_almost_equals
- Modified ALL assert methods to accept optional error message:
    assertEquals(actual, expected, "this if fails"),
    assertTable(t, "t had better be a table or else"), &c.;
    + assertError requires special mention:
      all other asserts follow built-in assert of message as last param;
      because of vararg in assertError, it handles things differently:
        assertError(f, ...) expects f to fail with the given args;
        assertError('err', f, ...) expects f to fail with the given args 
            and output the given message (err must be a string) ;
        assertError('msg', {'err' | nil}, f, ...) expects f to fail 
            and err (if provided) to be the error message generated 
                (must be string OR nil) ;
            and msg will be output if method does not fail
                or generated message does not match 'err' (if provided);
    see test_assert for more information and many examples.
- Renamed USE_EXPECTED_ACTUAL_IN ASSERT_EQUALS to USE_EXPECTED_ACTUAL.
- Added setter method to hide inner variable and check type:
    LuaUnit:setExpectedActual(boolean)
- Added variable DELTA_TOLERANCE for the default delta in assertAlmostEquals.
- Addded setter getter methods for:
    VERBOSITY       LuaUnit:getVerbosity()      LuaUnit:setVerbosity(lvl)
    DELTA_TOLERANCE LuaUnit:getDeltaTolerance() LuaUnit:setDeltaTolerance(n)
- Now properly accepts test targets via command-line,
    e.g., `lua test_file.lua TestClass1, TestClass2`.
- Refactored out separate modules with routines that are generally useful:
    moretables (table-printing; table-comparison)
    stringbuffer (string buffers) 
    stringsplit (split a string into a list around delimiters)
    asserts (semi-lightweight stand-alone library for assert patterns)
- Moved changelog from luaunit.lua to new file, CHANGELOG.txt, and
    tidied the version history.

[2.0]
- This is a major update that has some breaking changes to make it much more easy to use and code in many different styles
- Made the module only touch the global table for the asserts. You now use the module much more like Lua 5.2 when you require it.
  You need to store the LuaUnit table after you require it to allow you access to the LuaUnit methods and variables.
  (ex. local LuaUnit = require( "luaunit" ))
- Made changes to the style of which LuaUnit forced users to code there test classes. It now is more layed back and give the ability to code in a few styles.
    - Made "testable" classes able to start with 'test' or 'Test' for their name.
    - Made "testable" methods able to start with 'test' or 'Test' for their name.
    - Made testClass:setUp() methods able to be named with 'setUp' or 'Setup' or 'setup'.
    - Made testClass:tearDown() methods able to be named with 'tearDown' or 'TearDown' or 'teardown'.
    - Made LuaUnit.wrapFunctions() function able to be called with 'wrapFunctions' or 'WrapFunctions' or 'wrap_functions'.
    - Made LuaUnit:run() method able to be called with 'run' or 'Run'.
- Added the ability to tell if tables are equal using assertEquals. This uses a deep compare, not just the equality that they are the same memory address.
- Added LuaUnit.is<Type> and LuaUnit.is_<type> helper functions. (e.g. assert( LuaUnit.isString( getString() ) )
- Added assert<Type> and assert_<type> 
- Added assertNot<Type> and assert_not_<type>
- Added _VERSION variable to hold the LuaUnit version
- Added LuaUnit:setVerbosity(lvl) method to the LuaUnit table to allow you to control the verbosity now. If lvl is greater than 1 it will give verbose output.
  This can be called from alias of LuaUnit.SetVerbosity() and LuaUnit:set_verbosity().
- Moved wrapFunctions to the LuaUnit module table (e.g. local LuaUnit = require( "luaunit" ); LuaUnit.wrapFunctions( ... ) )
- Fixed the verbosity to actually format in a way that is closer to other unit testing frameworks I have used.
  NOTE: This is not the only way, I just thought the old output was way to verbose and duplicated the errors.
- Made the errors only show in the "test report" section (at the end of the run)

[1.3]
- port to lua 5.1
- use orderedPairs() to iterate over a table in the right order
- change the order of expected, actual in assertEquals() and the default value of
  USE_EXPECTED_ACTUAL_IN_ASSERT_EQUALS. This can be adjusted with
  USE_EXPECTED_ACTUAL_IN_ASSERT_EQUALS.

[1.2a]
- fix: test classes were not run in the right order

[1.2]
- tests are now run in alphabetical order
- fix a bug that would prevent all tests from being run

[1.1]
- internal variables are not global anymore
- you can choose between assertEquals( actual, expected) or assertEquals(
  expected, actual )
- you can assert for an error: assertError( f, a, b ) will assert that calling
  the function f(a,b) generates an error
- display the calling stack when an error is spotted
- a dedicated class collects and displays the result, to provide easy
  customisation
- two verbosity level, like in python unittest

[1.0]
- initial release



test_luaunit.lua
---Description: Tests for the luaunit testing framework
--
-- Built upon work by Philippe Fremy <phil@freehackers.org> and others
--
-- Author: J. 'KwirkyJ' Smith <kwirkyj.smith0@gmail.com>
-- Version: 3.0
-- License: X11 License

local LuaUnit    = require 'luaunit'
local moretables = require 'moretables'



---- verify that assertive library is loaded to global (_G) namespace --------

TestGlobalAsserts = {}
TestGlobalAsserts.testGlobal = function(self)
    assertEquals(3,3)
    assertEquals({3, {a=1}, true}, {3, {a=1}, true})
    assertNotEquals({4,3,{2}}, {4,3,{1}}, "nested element mismatch")
    assertString("this is a string!")
    assert_not_function(nil)
    assert_not_equals(true, 'green')
end



---- Class to show that tests are run in alphabetic order by default ---------

TestOrder = {}
    function TestOrder:setUp() end
    function TestOrder:tearDown() end
    function TestOrder:test1() end
    function TestOrder:testb() end
    function TestOrder:test3() end
    function TestOrder:testa() end
    function TestOrder:test5() end
    function TestOrder:test4() end
    function TestOrder:test2() end



---- TEST UTILITY FUNCTIONS (QA) ---------------------------------------------

TestWrapValue = {}
TestWrapValue.test_nil = function(self)
    assertEquals(LuaUnit._wrapValue(), 'nil')
end
TestWrapValue.test_number = function(self)
    assertEquals(LuaUnit._wrapValue(5), '5')
end
TestWrapValue.test_string = function(self)
    assertEquals(LuaUnit._wrapValue('5'), [=['5']=])
end
TestWrapValue.test_function = function(self)
    local f = function() end
    assertString(LuaUnit._wrapValue(f), tostring(f)) -- some function 0xNNNNN
end
TestWrapValue.test_boolean = function(self)
    assertEquals(LuaUnit._wrapValue(false), 'false')
end
TestWrapValue.test_table = function(self)
    local t = {[3]='b'}
    assertEquals(LuaUnit._wrapValue(t), moretables.tostring(t))
end

TestTrimErrMsg = {}
TestTrimErrMsg.test_empty = function(self)
    assertEquals(LuaUnit._trimErrMsg(''), '')
end
TestTrimErrMsg.test_boring_string = function(self) 
    assertEquals(LuaUnit._trimErrMsg(
                    "string that doesn't match the pattern at all"),
                 "string that doesn't match the pattern at all")
end
TestTrimErrMsg.test_no_line_numbers = function(self)
    assertEquals(LuaUnit._trimErrMsg(
                    "faulty:string: misses line number"),
                 "faulty:string: misses line number")
end
TestTrimErrMsg.test_whitespace_trimmed = function(self)
    assertEquals(LuaUnit._trimErrMsg("\ttrimmed whitespace!\n"),
                 "trimmed whitespace!") 
end
TestTrimErrMsg.test_successful = function(self)
    assertEquals(LuaUnit._trimErrMsg(
                    'someth1ng_thing.ext:123: message thing\ncontinued\n'),
                 'message thing\ncontinued') 
end
TestTrimErrMsg.test_non_string = function(self)
    assertError(LuaUnit._trimErrMsg, 5)
end

--TODO: verbosity manipulation
--TODO: delta manipulation
--TODO: mechanism to fetch/redirect output
--TODO: expected-actual / actual-expected toggle

-- LuaUnit:setVerbosity(0) -- output is much less
-- LuaUnit:setDeltaTolerance(1e-10) -- set default 'almost-equals' delta

-- LuaUnit:run('TestToto:test3') -- TODO: will execute only one test
-- LuaUnit:run('TestToto') -- will execute only one test class
-- LuaUnit:run() -- run all tests in tables starting with 'Test' or 'test'
LuaUnit:run(arg) -- like above, but can select classes via command line
--    e.g.,  `lua test_file.lua TestUtil TestLuaUnit:test_assertNotEquals`



test_classrunner.lua
-- unit-test for the classrunner

local ClassRunner = require 'classrunner'
local LuaUnit     = require 'luaunit'

--TODO: test the (self) variable in test classes

---- STUFF -------------------------------------------------------------------

local a,b,c,d = 0,0,0,0 -- variables to show call frequency
UnTestTiti = {
    setUpClass = function(self)
        a=a+1
    end,
    setUp = function(self)
        b=b+1
    end,
    tearDown = function(self)
        c=c+1
    end,
    tearDownClass = function(self)
        d=d+1
    end,
    test1 = function(self) end,
    testb = function(self) end,
    test3 = function(self) end,
}

function some_function() end

function fail_function() assert(false) end



---- TEST CLASSRUNNER  -------------------------------------------------------

TestClassRunner = {}
TestClassRunner.test_auto_name = function(self)
    local cr
    for i=1,6 do
        cr = ClassRunner.new()
        assertEquals(cr:getName(), 'ClassRunner'..i)
    end
end
TestClassRunner.test_nominal = function(self)
    local cr = ClassRunner.new('UnTestTiti')
    cr:setSilent(true) -- true is redundant
    assertEquals(cr:getOutput(), '')
    cr:run()
    assertEquals({a,b,c,d}, {1,3,3,1}, 
                 'call frequency of setups/teardowns')
    assertEquals({cr:getResults()}, {3, {}, 0},
                 'three successes, no failures, no skips')
    assertEquals(cr:getOutput(), 
                 'UnTestTiti\n  [test1]\tOk\n  [test3]\tOk\n  [testb]\tOk\n',
                 'alphabetic test order, verbose output')
end
TestClassRunner.test_specified = function(self)
    local cr = ClassRunner.new('UnTestTiti')
    cr:addTest('fail_function')
    cr:setSilent()
    --note order; t2 will be absent; test1 will not be run twice
    cr:run('testb', 'test1', 't2', 'test1', 'fail_function')
    local ran, fails, skips = cr:getResults()
    assertEquals(ran, 3)
    assertEquals(#fails, 1)
    assertEquals(#fails[1], 3)
    local name, msg, stack = fails[1][1], fails[1][2], fails[1][3]
    assertEquals(name, 'fail_function')
    assertEquals(msg, 'test_classrunner.lua:31: assertion failed!')
    assertString(stack, 'stack traceback expected here')
    assertEquals(skips, 1)
    assertEquals(cr:getOutput(), 
                 'UnTestTiti\n  [testb]\tOk\n  [test1]\tOk\n  '..
                 '[t2]\tAbsent\n  [fail_function]\tFailed\n')
end
TestClassRunner.test_build_class = function(self)
    local cr = ClassRunner.new('MyTestClass')
    cr:addTest('some_function')
    cr:setSilent()
    cr:run()
    assertEquals(cr:getOutput(), 'MyTestClass\n  [some_function]\tOk\n')
end
TestClassRunner.test_rerun_resets = function(self)
    local cr = ClassRunner.new('UnTestTiti')
    cr:setSilent()
    cr:run()
    assertEquals( {cr:getResults()}, {3, {}, 0})
    assertEquals(cr:getOutput(),
                 'UnTestTiti\n  [test1]\tOk\n  [test3]\tOk\n  [testb]\tOk\n')
    cr:addTest('some_function')
    cr:run('some_function', 'testb')
    assertEquals( {cr:getResults()}, {2, {}, 0}) -- ran some_function and testb
    assertEquals(cr:getOutput(), 
                 'UnTestTiti\n  [some_function]\tOk\n  [testb]\tOk\n')
end



---- TEST VERBOSITY ----------------------------------------------------------

TestVerbosity = {}
TestVerbosity.setUp = function(self)
    -- also re-demonstrate building a test class 'from scratch'
    self.cr = ClassRunner.new('TestAbsentClass')
    self.cr:addTest('some_function')
    self.cr:addTest('fail_function')
    self.cr:setSilent() -- do not print while running
end
TestVerbosity.test_terse_set = function(self)
    self.cr:setVerbosity(0)
    self.cr:run()
    local total, fails = self.cr:getResults()
    assertEquals(total, 2)
    assertEquals(self.cr:getOutput(), 'TestAbsentClass\tF.\n')
    local name, msg, stack = fails[1][1], fails[1][2], fails[1][3]
    assertEquals(#fails, 1)
    assertEquals(name, 'fail_function', 
                 'failing function name')
    assertEquals(msg, 'test_classrunner.lua:31: assertion failed!',
                 'failing function message')
    assertNotNil(stack:find('in function'), 
                 'peek at stacktrace')
end
TestVerbosity.test_terse_param = function(self)
    local runnerWithParam = ClassRunner.new('TestVerbParam', 0)
    runnerWithParam:addTest('some_function')
    runnerWithParam:addTest('fail_function')
    runnerWithParam:setSilent()
    runnerWithParam:run()
    assertEquals(runnerWithParam:getOutput(), 'TestVerbParam\tF.\n')
    local total, fails = runnerWithParam:getResults()
    local name, msg, stack = fails[1][1], fails[1][2], fails[1][3]
    assertEquals(#fails, 1)
    assertEquals(name, 'fail_function', 
                 'failing function name')
    assertEquals(msg, 'test_classrunner.lua:31: assertion failed!',
                 'failing function message')
    assertNotNil(stack:find('in function'), 
                 'peek at stacktrace')
end
TestVerbosity.test_verbose = function(self)
    assertEquals(self.cr:getVerbosity(), 1) -- default verbosity
    self.cr:run()
    local total, fails = self.cr:getResults()
    assertEquals(total, 2)
    assertEquals(self.cr:getOutput(),
[[TestAbsentClass
  [fail_function]	Failed
  [some_function]	Ok
]])
    assertNotNil(fails[1][3]:find('in function'))
end
TestVerbosity.test_module_set = function(self)
    local default, runner = ClassRunner.getVerbosity(), nil
    assertEquals(default, 1)
    for i = 0, 5 do
        ClassRunner.setVerbosity(i)
        runner = ClassRunner.new('TestClass')
        assertEquals(runner:getVerbosity(), i)
    end
    ClassRunner.setVerbosity(nil)
    assertEquals(ClassRunner.getVerbosity(), default)
end



LuaUnit:run(arg)



lib/stringsplit/test/test_stringsplit.lua
-- Unit-test and use-documentation for stringsplit.

local split = require('stringsplit')['split']

local function alike(t1, t2)
    assert(type(t1) == 'table' and type(t2) == 'table',
           'inputs must be tables')
    if #t1 ~= #t2 then return false end -- tables must be of same length
    for i=1, #t1 do
        if t1[i] ~= t2[i] then
            return false
        end
    end
    return true
end



local fox = 'the quick brown fox jumped over the lazy dog'



---- STANDARD BEHAVIOR ----------------------------------------------------

assert(alike(split("o", fox),
             {'the quick br', 'wn f', 'x jumped ', 'ver the lazy d', 'g'}),
       'standard behavior')

assert(alike(split("p", 'pizza'), {'', 'izza'}),
       'match at start of string')

assert(alike(split('g', fox), 
             {'the quick brown fox jumped over the lazy do', ''}),
       'match at end of string')

assert(alike(split("[k-s]", 'pizza'), {'', 'izza'}),
       'pattern-match at start of string')

assert(alike(split(",%s*", "Anna, Bob, Charlie,Dolores"),
             {'Anna', 'Bob', 'Charlie', 'Dolores'}),
       'pattern-match')

assert(alike(split(",", fox), {fox}),
       'no match')

assert(alike(split('a', ''), {''}),
       'empty string')



---- ERROR CONDITIONS ------------------------------------------------------

---Try to trim a string like 'test_file.lua:125: assertion failed!\n' 
--                        to 'assertion failed!'
-- @param s {String}
-- @error Iff s is not a {String}.
-- @return {String}
local function stripErrMsgHeader(s)
    assert (type(s) == 'string', 's must be a string!')
    s = s:gsub('^%s*(.-)%s*$', '%1') -- remove whitspace
    return s:gsub('.*:%d+: (.-)', '%1') -- assuming ':%d+: ' is the line num
end

local delimiters = { '', nil,   5, 'j', 'j'}
local strings    = {fox, fox, fox, nil,   5}
local messages = {'delimiter cannot be empty string', 
                  'delimiter must be a string', 
                  'delimiter must be a string',  
                  'text must be a string', 
                  'text must be a string'}
local ok, err
for i=1, #delimiters do
    ok, err = pcall(split, delimiters[i], strings[i])
    assert(not ok, 'assertion failed at index ' .. i)
    assert(stripErrMsgHeader(err) == messages[i],
           'err at: ' .. i .. ' : ' .. err)
end


print ('==== TEST_STRINGSPLIT PASSED ====')



lib/stringsplit/init.lua
-- Lightweight module implementing a routine to split a string.
--
-- sample usage:
-- local stringsplit = require 'stringsplit' -- stringsplit on path
-- split = stringsplit.split 
-- local s = split('t', 'earth')
-- assert (type(s) == 'table' and #s == 2)
-- assert (s[1] == 'ear' and s[2] == 'h')
--
-- Author: J. 'KwirkyJ' Smith <kwirkyj.smith0@gmail.com>
-- Year: 2016
-- Version: 1.0.0
-- License: MIT(X11) License



local stringsplit = {_VERSION = '1.0.0'}

---split(delimiter, text)
-- Split text into a list consisting of the strings in text
-- separated by the delimiter; 
-- example with Pattern: 
-- strsplit(",%s*", "Anna, Bob, Charlie,Dolores")
-- >> {'Anna', 'Bob', 'Charlie', 'Doloes'} Pattern eats whitespace.
-- @param delimiter {String} Any non-empty string, can be a Pattern.
-- @param text      {String} Any string, can be empty.
-- @error iff delimiter or text are not {String} type;
--        elseiff delimiter is the empty string ('').
-- @return {Table} Array of strings; if first character(s) match pattern
--          the first element will be ''; no match returns {<text>}.
stringsplit.split = function(delimiter, text)
    local list = {}
    local pos = 1
    assert (type(delimiter) == 'string', 'delimiter must be a string')
    assert (type(text) == 'string', 'text must be a string')
    assert (delimiter ~= '', 'delimiter cannot be empty string')
    local start, stop
    while 1 do
        start, stop = string.find(text, delimiter, pos)
        if start then
            table.insert(list, string.sub(text, pos, start-1))
            pos = stop+1
        else
            table.insert(list, string.sub(text, pos))
            break
        end
    end
    return list
end

return stringsplit



luaunit.lua
-- 
--         luaunit.lua
--
-- Description: A unit testing framework
-- Homepage: http://phil.freehackers.org/luaunit/
-- Initial author: Ryu, Gwang (http://www.gpgstudy.com/gpgiki/LuaUnit)
-- Lot of improvements by Philippe Fremy <phil@freehackers.org>
-- More improvements by Ryan P. <rjpcomputing@gmail.com>
-- Further improvements by J. 'KwirkyJ' Smith <kwirkyj.smith0@gmail.com>
-- Version: 3.0
-- License: X11 License, see LICENSE.txt

---- SETUP -------------------------------------------------------------------

-- lua 5.2+ deprecates unpack in favor of table.unpack
unpack = table.unpack or unpack

-- lua 5.2+ deprecates/removes loadstring() for load()
loadstring = loadstring or load

-- Some people like assertEquals( actual, expected ) 
-- and some people prefer assertEquals( expected, actual ).
-- Set to false for the former.
local USE_EXPECTED_ACTUAL = false
local DELTA_TOLERANCE = 1e-12
local VERBOSITY = 1

local classrunner  = require 'luaunit.classrunner'
local assertive    = require 'assertive'
--local StringBuffer = require 'lua_stringbuffer'
local toString = require('moretables')['tostring']
local strsplit = require('luaunit.lib.stringsplit')['split']

---- HELPER FUNCTIONS --------------------------------------------------------

---Wrapper for tostring to differentiate string types.
-- @param v Value to convert to string.
-- @return {String} tostring(v) iff not already a string;
--                  else '<v>'.
local function wrapValue(v)
    if type(v) == 'string' then return "'"..v.."'" end
    return toString(v)
end

--[[
-- Order of testing
local function __genOrderedIndex( t )
    local orderedIndex = {}
    for key,_ in pairs(t) do
        table.insert( orderedIndex, key )
    end
    
    -- assumption: only numbers and strings can be keys (values in orderedI)
    -- if same type, use '<'
    -- else, numbers come first
    local function comp(a,b)
        if type(a) == type(b) then
            return a < b
        elseif type(a) == 'number' then
            return true
        else return false
        end
    end
    
    table.sort( orderedIndex, comp )
    return orderedIndex
end

---Equivalent of the next() function of table iteration, but returns the
-- keys in the alphabetic order. We use a temporary ordered key table that
-- is stored in the table being iterated.
local function orderedNext(t, state)

    --print("orderedNext: state = "..tostring(state) )
    if state == nil then
        -- the first time, generate the index
        t.__orderedIndex = __genOrderedIndex( t )
        local key = t.__orderedIndex[1]
        return key, t[key]
    end
    -- fetch the next value
    local key = nil
    for i = 1,#t.__orderedIndex do
        if t.__orderedIndex[i] == state then
            key = t.__orderedIndex[i+1]
        end
    end

    if key then
        return key, t[key]
    end

    -- no more value to return, cleanup
    t.__orderedIndex = nil
    return
end

---Iterator function to go over a table in predictable sequence (alphabetic) ;
-- called in the same manner as pairs().
orderedPairs = function(t) -- filling forward definition; is local
    return orderedNext, t, nil
end
--]]

--[[
---Removes some information from stacktrace to be relevant to the tests.
local function strip_luaunit_stack(stack_trace)
    local stack_list = strsplit( "\n", stack_trace )
    local strip_end = nil
    for i = #stack_list,1,-1 do
        -- a bit rude but it works !
        if string.find(stack_list[i],"[C]: in function `xpcall'",0,true)
            then
            strip_end = i - 2
        end
    end
    if strip_end then
        table.setn( stack_list, strip_end )
    end
    local stack_trace = table.concat( stack_list, "\n" )
    return stack_trace
end
--]]

---Trim a string like 
-- 'test_file.lua:125: assertion failed!\n' to
-- 'assertion failed!'
-- @param s {String}
-- @error Raised if s is not a {String}.
-- @return {String}
local function trimErrMsg(s)
    assert (type(s) == 'string', 's must be a string!')
    s = s:gsub('^%s*(.-)%s*$', '%1') -- remove whitspace
    return s:gsub('.*:%d+: (.-)', '%1') -- assuming ':%d+: ' is the line num
end



---- ASSERT ROUTINES ---------------------------------------------------------

---Register all the asserts to the global namespace with both
-- camelCase and underscore_lowercase options:
-- e.g., assertError(f, ...) and assert_error(f, ...).
-- assertAlmostEquals
-- assertError
-- assertEquals
-- assertNotEquals
-- assert<Type> (for all standard types)
-- assertNot<Type> (for all standard types)
for k,v in pairs(assertive) do
    if type(k) == 'string'
    and string.sub(k, 1, 6) == 'assert'
    then
        _G[k] = v
        _G[k:gsub('(%u)', '_%1'):lower()] = v
    end
end



---- LUAUNIT CLASS -----------------------------------------------------------

local LuaUnit = { _VERSION = "3.0.0" }

--[[
---Get the current verbosity level.
-- LuaUnit:getVerbosity()
--@return {Number} >= 0.
local function getVerbosity(self)
    return VERBOSITY
end
LuaUnit.getVerbosity  = getVerbosity
LuaUnit.get_verbosity = getVerbosity

---Set the verbosity of output.
-- LuaUnit:setVerbosity(1)
-- @param lvl {number} If > 0 there will be verbose output (default of 0).
local function setVerbosity(self, lvl)
    lvl = lvl or 0
    assert(type(lvl) == 'number', 'Verbosity must be a number')
    self.result.verbosity = lvl
    VERBOSITY = lvl
    assertive:setVerbosity(lvl)
end
LuaUnit.setVerbosity  = setVerbosity
LuaUnit.set_verbosity = setVerbosity
LuaUnit.SetVerbosity  = setVerbosity
--]]

--[[
---Get the current verbosity level.
-- LuaUnit:getDeltaTolerance()
--@return {Number} >= 0.
local function getDeltaTolerance(self)
    return DELTA_TOLERANCE
end
LuaUnit.getDeltaTolerance   = getDeltaTolerance
LuaUnit.get_delta_tolerance = getDeltaTolerance
    
---Set the default maximum delta in assertAlmostEquals.
-- LuaUnit:setDefaultTolerance(1e-7)
-- @param n {Number} (default of 1e-12).
local function setDeltaTolerance(self, n)
    n = n or 1e-12
    assert(type(DELTA_TOLERANCE) == 'number', 'must be number')
    DELTA_TOLERANCE = n
    assertive:setDelta(n)
end
LuaUnit.setDeltaTolerance   = setDeltaTolerance
LuaUnit.set_delta_tolerance = setDeltaTolerance
--]]

--[[
---Set the inner variable to configure param order in assert[Not]Equals.
-- LuaUnit:setExpectedActual(true)
-- @param b {Boolean} (default of true).
local function setExpectedActual(self, b)
    -- I know self is unused, and YOU know self is unused,
    -- but the 'self' variable makes the 'self-modifying colon' sensible.
    if type(b) ~= 'boolean' then b = true end
    b = b
end
LuaUnit.setExpectedActual   = setExpectedActual
LuaUnit.set_expected_actual = setExpectedActual
--]]



local function runAllTests(runners)
    local runsum, failsum, byesum, failslist = 0, 0, 0, {}
    for _, runner in ipairs(runners) do
        runner:run()
        local runcount, fails, byecount = runner:getResults()
        runsum = runsum + runcount
        failsum = failsum + #fails
        byesum = byesum + byecount
        if #fails > 0 then
            failslist[#failslist+1] = {runner:getName(), fails}
        end
    end
    return runsum, failsum, byesum, failslist
end

local function printFailures(faillist, verbosity)
    if #faillist == 0 then return end
    print('=========================================================')
    print('Failed tests:\n-------------')
    local classname, fails, methodname, message, stack
    for i=1, #faillist do
        classname, fails = faillist[i][1], faillist[i][2]
        for j=1, #fails do
            methodname, message, stack = fails[j][1], fails[j][2], fails[j][3]
            print('>>> ' .. classname .. ':' .. methodname)
            print(message)
            print('stack traceback:')
            print(stack)
        end
    end
end

local function printFinalSummary(total, failed, byes)
    local s, ratio
    print('=========================================================')
    if total == 0 or failed == 0 then
        ratio = 1
    else 
        ratio = ((total-failed) / total)
    end
    s = string.format('Success: %0.0f%%\t(%d/%d)',
                      100*ratio, total-failed, total)
    if byes > 0 then s = string.format('%s\t(skipped: %d)', s, byes) end
    print(s)
end



---Run a test suite.
-- LuaUnit:run([{arg | ...}])
-- LuaUnit:run('TestClass', 'SomeTestClass', ...)
-- @param arg {Table} Optional (recommended) parameter to pass command-
--            line arguments; used to specify test classes:
--            in test_file.lua: `LuaUnit:run(arg)`  
--            in shell:         `$ lua test_file.lua TestClass1 TestClass2`
-- @param ... {String[, String]} Optional Sequence of strings that match 
--            'class names' in your test file, e.g. TestToto1;
--           TODO: alternatively, Class:Method names, e.g. TestToto1:test_kansas
LuaUnit.run = function(self, ...)
    local args, runners = {...}, {}
    if args[1] and type(args[1]) == 'table' then
        args = args[1] -- assumes  is the passed-through arg table
    end
    if #args > 0 then
        for _,v in ipairs(args) do
            if type(_G[v]) == 'table' then
                runners[#runners + 1] = classrunner.new(v)
            end
        end
    else
        for k,v in pairs(_G) do 
            if  type(v) == "table" 
            and (string.sub(k, 1, 4) == "Test" or 
                 string.sub(k, 1, 4) == "test")
            then
                runners[#runners + 1] = classrunner.new(k)
            end
        end
    end
    local runsum, failsum, byesum, failslist = runAllTests(runners)
    printFailures(failslist)
    printFinalSummary(runsum, failsum, byesum)
end
LuaUnit.Run = LuaUnit.run -- alias



-- register below routines to verify correctness
--LuaUnit._strsplit = strsplit
LuaUnit._toString = toString
LuaUnit._wrapValue = wrapValue
LuaUnit._trimErrMsg = trimErrMsg


return LuaUnit

